content/languages/c/authoring/index.md

```diff
@@ -0,0 +1,285 @@
+- Some header files include other header files indirectly, for example, file `foo.h` contains line `#include <bar.h>`, which might appear to make the explicit include for `bar.h` unnecessary. It's not true though, because the file `foo.h` might change one day, or might depend on some compiler settings or command line options, and after some changes to the configuration of the C runner, the `bar.h` might be not included there anymore. That's why every file (i.e. code snippet) of a kata should explicitly include all required header files declaring functions used in it.
+- The author might think that header files for the testing framework are included automatically by the code runner. That is not the case though, and test suites need to include `criterion/criterion.h` explicitly.
+
+### Compilation warnings
+
+Compiler options related to warnings used by the C runner are somewhat strict and require some discipline to get the code to compile cleanly. `-Wall` and `-Wextra` may cause numerous warnings and some of them are very pedantic. However, code of C kata should still compile cleanly, without any warnings logged to the console. Even when a warning does not cause any problem with tests, users get distracted by them and blame them for failing tests.
+
+
+### Memory management
+
+Unlike many modern, high-level languages, C does not manage memory automatically. Manual memory management is a very vast and complex topic, with many possible ways of achieving the goal depending on a specific case, caveats, and pitfalls.
+
+Whenever a kata needs to return a string or an array, C authors tend to use the naive technique of allocating the memory in the solution function, and freeing it in the test suite. This approach mimics the behavior known from other languages where returning an array or object from inside of the user's solution is perfectly valid, but it's hardly ever a valid way of working with unmanaged memory.
+
+One of the consequences of unmanaged memory is that it's strongly recommended against returning string constants from C functions, especially when translating kata from other languages. Returning a string in other languages is not a problem, but in C it always raises questions of who should allocate it and how it should be allocated. Consider replacing the string with some simpler data type (eventually aliased with a `typedef`), and/or provide some symbolic constants for available values. For example, if the requirement for the Python version is to _Return the string `'You won!'` if you won the game, `'You lost :('` if you were defeated, and `'It's a draw'` if there is no winner._, C should preferably use the constants `WON`, `LOST` and `DRAW`. If the author decides to keep raw C-strings as elements of the kata interface, they should clearly specify the required allocation scheme.
```

Comment:
`'You won!'` isn't just an enum-like `'won'` (or should it be `'win'` for consistency with `'draw'`?..), it's a formatted string to be shown, so it kind of is consistent with returning a `const char*`, but creating a user-friendly string isn't really something that the function should do. So the type probably should be changed, but it isn't because C is different somehow.

Reply:
Please see [this comment](https://github.com/codewars/docs/pull/228#discussion_r554881117) and [this comment](https://github.com/codewars/docs/pull/228#discussion_r554930522). Maybe my example was not really fortunate, because I specifically meant kata with interface like, for example, "Red Knight", which says "return the color of captured pawn" while only "White" and "Black" are possible, or "Connect Four" with return values of "Plyer_Red" and "Player_Green" etc. I will think how to improve the example.

Reply:
I tried to change the example so it would stress the point better. Do you think it's better now?